### PR TITLE
`zombienet`: warp-sync tests for validators 

### PR DESCRIPTION
**Pull request by michalkucharczyk**
Sketch of warp-sync test for validators. ~~Not tested.~~

~~`warp-sync` is now failing with `babe-skip-epochs`, the reason is here: https://github.com/paritytech/substrate/pull/12751#issuecomment-1328032781~~

Summary of tests (copied from comment):
1. [_0001_](https://github.com/paritytech/substrate/blob/mku-validators-warp-sync-test/zombienet/0001-basic-warp-sync/test-warp-sync.toml): full nodes only: three synced from db, one warp-synced,
2. [_0002_](https://github.com/paritytech/substrate/blob/mku-validators-warp-sync-test/zombienet/0002-validators-warp-sync/test-validators-warp-sync.toml): two validators warp-synced,  three full nodes synced from db
3. [_0003_](https://github.com/paritytech/substrate/blob/mku-validators-warp-sync-test/zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.toml): validators synced from db, one full node synced from db, full node warp-synced

Closes: #12769

### CODE REVIEW

**Comment by davxy** [comment]
Dumb question. Can't we merge this test (0002) with 0001 by warp syncing two nodes there?
I.e. by adding a warp sync for a validator there?

**Comment by michalkucharczyk** [comment]
> Dumb question. Can't we merge this test (0002) with 0001 by warp syncing two nodes there? I.e. by adding a warp sync for a validator in 0001?

For me those scenarios are slightly different, and different failures are observed. I would prefer to keep them separate. Just to summarize:

1. 0001: full nodes only: three synced from db, one warp-synced,
2. 0002: two validators warp-synced,  three full nodes synced from db
3. 0003: validators synced from db, one full node synced from db, full node warp-synced

**Comment by davxy** [comment]
Yeah. Your are right, better to separate the scenarios.

---

So as we've discussed in the chat we have 2 issues to resolve.
1. (test case 0002) : validator **fails** after being warp synced
2. (test case 0003) : full node **fails** after being warp synced

Both the cases are due to epoch skipping feature not handling two edge cases.

Case 3 has been already deeply discussed https://github.com/paritytech/substrate/pull/12751#issuecomment-1328032781
and we need a fix for the `gap`. I may attempt to relax the gap checks when searching for epoch data.

---

Case 1.

**TL;DR: the issue here is that the oracle `is_major_syncing()` returns false (sync state is Idle) but the blocks were not all imported yet by the backend **

Long explanation:

Here the failure is due to a validator trying to build a new block after being warp synced.

This is the execution stack triggering the failure (less depth first)

- `next_slot()` is called by the block authoring worker when is time to attempt block production
- `create_inherent_data_provider()` closure called with the last best block (i.e. the one the sync leaves us with) [here](https://github.com/paritytech/substrate/blob/4692572ecf8de0f9420972f8bf2f072243cf5a94/client/consensus/slots/src/slots.rs#L159-L162)
- `create_uncles_inherent_data_provider()` called [here](https://github.com/paritytech/substrate/blob/4692572ecf8de0f9420972f8bf2f072243cf5a94/bin/node/cli/src/service.rs#L438-L441)
- `Client::uncles()` called [here](https://github.com/paritytech/substrate/blob/4692572ecf8de0f9420972f8bf2f072243cf5a94/client/consensus/uncles/src/lib.rs#L42)
- ...
- Starting from current best block (0x5448..., i.e. the point where the warp sync positioned us)
  we begin collecting uncles [here](https://github.com/paritytech/substrate/blob/4692572ecf8de0f9420972f8bf2f072243cf5a94/client/service/src/client/client.rs#L1093-L1108).
  - (hash=0x5445..., height=12133) => OK   (THIS IS THE BEST BLOCK according to warp sync)
  - (hash=0x9e94..., height=12132) => OK
  - (hash=0xaea8..., height=12131) => OK   (THIS IS THE LAST FINALIZED BLOCK according to warp sync)
  - (hash=0x2e3b..., height=12130) => FAIL (block header not found error)
   
That is a bit strange since when the oracle returns `is_major_syncing() == false` the headers
should have been downloaded (i.e. we don't attempt block authoring if we've not downloaded the blocks)

So [here](https://github.com/paritytech/substrate/blob/4692572ecf8de0f9420972f8bf2f072243cf5a94/client/consensus/slots/src/lib.rs#L530-L531)
instead of `return` I've inserted a `continue`.

And indeed it fails 2 times and then at the third attempt it works.

If we print the logs of babe we see that when the `next_slot()` is invoked the `block_import` for old blocks has not finished to
import **all** the blocks yet. 

When all the blocks are imported, 0x2e3b is now present.

(Note that the blocks from last finalized (12131) to best (12133) are imported first... then the import begins from 1, 2 ,3 ...)

To wrap-up, looks like in the `next_slot()` it is not sufficient to check if the network has finished downloading the blocks. We also need to check if we imported the one below the last finalized.

File: zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl [hunk 1]
```diff
@@ -0,0 +1,43 @@
+Description: Warp sync
+Network: ./test-validators-warp-sync.toml
+Creds: config
+
+alice: is up within 30 seconds
```
Comment (melekes) [inline zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl:5]: this is so cool ❤️

File: zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.toml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  name = "alice"
+  validator = true
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
+
+  [[relaychain.nodes]]
+  name = "bob"
+  validator = true
+  db_snapshot="https://storage.googleapis.com/zombienet-db-snaps/substrate/0001-basic-warp-sync/chains-0bb3f0be2ce41b5615b224215bcc8363aa0416a6.tgz"
+
+  #we need at least 3 nodes for warp sync
```
Comment (melekes) [inline zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.toml:22]: suggestion: either move up or delete this comment. In the above `.toml` file there are 3 descriptions after the same comment, but, here, there are only two.

Comment (michalkucharczyk) [reply zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.toml:22]: Yep, it may be confusing. Actually this comment relates to third node, explaining why we need it.

**Comment by davxy** [comment]
Test case 3 (https://github.com/paritytech/substrate/pull/13078#issuecomment-1374021848) should be fixed by https://github.com/paritytech/substrate/pull/13106

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by davxy** [comment]
Test case 2 (https://github.com/paritytech/substrate/pull/13078#issuecomment-1374021848) should be fixed by https://github.com/paritytech/substrate/pull/13155

**Review by davxy — APPROVED** [review]
LGTM, but lets wait for https://github.com/paritytech/substrate/pull/13155 to me merged on master and then rebase

**Review by melekes — APPROVED** [review]
👍

**Comment by michalkucharczyk** [comment]
> That is a bit strange since when the oracle returns `is_major_syncing() == false` the headers
> should have been downloaded (i.e. we don't attempt block authoring if we've not downloaded the blocks)

Dumb question: why cannot `is_major_syncing` be changed to reflect the state of import?

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by davxy** [comment]
Requires https://github.com/paritytech/substrate/pull/13135

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
Test are currently not passing due to: https://github.com/paritytech/zombienet/issues/700

File: zombienet/0002-validators-warp-sync/test-validators-warp-sync.toml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+[settings]
+enable_tracing = false
```
Comment (bkchr) [inline zombienet/0002-validators-warp-sync/test-validators-warp-sync.toml:2]: What is this doing?

File: zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+# workaround for: https://github.com/paritytech/zombienet/issues/580
+alice: count of log lines containing  "Block history download is complete" is 1 within 60 seconds
+bob: count of log lines containing  "Block history download is complete" is 1 within 60 seconds
+
+alice: reports block height is at least 12133 within 10 seconds
+bob: reports block height is at least 12133 within 10 seconds
+
+alice: count of log lines containing "error" is 0 within 10 seconds
+bob: count of log lines containing "verification failed" is 0 within 10 seconds
```
Comment (bkchr) [inline zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl:39]: Do we really need this?

File: zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+alice: reports peers count is at least 3 within 60 seconds
+bob: reports peers count is at least 3 within 60 seconds
+charlie: reports peers count is at least 3 within 60 seconds
+dave: reports peers count is at least 3 within 60 seconds
+
+
+# db snapshot has 12133 blocks
+dave: reports block height is at least 1 within 60 seconds
+dave: reports block height is at least 12132 within 60 seconds
```
Comment (bkchr) [inline zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl:23]: Is this the block we warp sync to? Or why checking for this particular block?

File: zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl [hunk 6]
```diff
@@ -0,0 +1,35 @@
+Description: Warp sync
```
Comment (bkchr) [inline zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl:1]: I'm a little bit confused why we have this extra test. Could we not at least merge them with the other one?

**Review by alvicsam — APPROVED** [review]

File: zombienet/0002-validators-warp-sync/test-validators-warp-sync.toml (hunk 3 above)
Comment (michalkucharczyk) [reply zombienet/0002-validators-warp-sync/test-validators-warp-sync.toml:2]: https://github.com/paritytech/zombienet/pull/582

File: zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl (hunk 4 above)
Comment (michalkucharczyk) [reply zombienet/0002-validators-warp-sync/test-validators-warp-sync.zndsl:39]: all bugs till now manifests with this log, so it is a way to check all is ok.

File: zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl (hunk 5 above)
Comment (michalkucharczyk) [reply zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl:23]: db has 12133 blocks.

File: zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl (hunk 6 above)
Comment (michalkucharczyk) [reply zombienet/0003-block-building-warp-sync/test-block-building-warp-sync.zndsl:1]: We could, but I’d prefer to keep them separate, they are slightly different. I think having smaller tests is simpler in terms of debugging.

Please refer to: https://github.com/paritytech/substrate/pull/13078#issuecomment-1373506149 for some details.

**Review by sergejparity — APPROVED** [review]
